### PR TITLE
Target turbo with turbo

### PR DESCRIPTION
**Pull request by ading1977**
This is the first pull request of target turbo with turbo, mainly addressing the following issues:

* Support application commodities such as Heap, Collection, Threads in addition to the existing Transaction and ResponseTime
* Support database server entities, and commodities such as DBMem, DBCacheHitRate, Connection, Transaction
* Update supply chain and stitching metadata accordingly
* Enhance `appmetric` configuration to allow user specify if an application or database server is hosted on VM. Remove the `createProxyVM` command line argument.
* Enhance `prometurbo` to create different providers/consumers based on `hostedOnVM` flag:
  * If `hostedOnVM` is `true`, then create:
    VM (proxy) -> Application/DBServer -> vApp -> BusinessApp
  * If `hostedOnVM` is `false` (default, implying hosted on container), then create:
    Application (proxy) -> vApp (proxy) -> BusinessApp 
* Allow default capacity of commodity (e.g., the default capacity of DBCacheHitRate and CollectionTime are 100% ), there are no queries needed/available for these capacities. 
* Update `turbo-go-sdk` and `vendor`

### CODE REVIEW

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go [hunk 1]
```diff
@@ -282,30 +245,26 @@ func (b *entityBuilder) createEntityDto(provider *proto.EntityDTO) (*proto.Entit
...
+	var commodities []*proto.CommodityDTO
+	var commTypes []proto.CommodityDTO_CommodityType
 
-		if _, ok := constant.AppCommodityTypeMap[commType]; !ok {
-			err := fmt.Errorf("unsupported commodity type %s", key)
-			glog.Errorf(err.Error())
+	for commType, value := range metric.Metrics {
+		defaultValue, ok := supportedCommodities[commType]
+		if !ok {
+			glog.Warningf("Unsupported commodity type %v for entity type %v", commType, entityType)
```
Comment (chlam4) [inline prometurbo/pkg/discovery/dtofactory/entity_builder.go]: Do I read it right that the default value is omitted in most of the commodities?  I'm sure I miss something so would just like to understand more...

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go [hunk 2]
```diff
@@ -282,30 +245,26 @@ func (b *entityBuilder) createEntityDto(provider *proto.EntityDTO) (*proto.Entit
...
 			continue
 		}
 
-		commodity, err := builder.NewCommodityDTOBuilder(commType).
-			Used(value[exporter.Used]).Key(commKey).Create()
+		commodityBuilder := builder.NewCommodityDTOBuilder(commType).
+			Used(value[exporter.Used]).Key(commKey)
```
Comment (chlam4) [inline prometurbo/pkg/discovery/dtofactory/entity_builder.go]: Nit: Are we assuming the used value is always there?  Do we need to check exists?

File: prometurbo/pkg/registration/supply_chain_factory.go [hunk 3]
```diff
@@ -103,26 +146,52 @@ func (f *SupplyChainFactory) CreateSupplyChain() ([]*proto.TemplateDTO, error) {
...
+	vmBuilder.SetPriority(-1)
+	vmBuilder.SetTemplateType(proto.TemplateDTO_BASE)
 
-	return builder.Create()
+	return vmBuilder.Create()
 }
 
+func (f *SupplyChainFactory) buildDBServerSupplyBuilder() (*proto.TemplateDTO, error) {
+	dbServerToVMExternalLink, err := supplychain.NewExternalEntityLinkBuilder().
+		Link(proto.EntityDTO_DATABASE_SERVER, proto.EntityDTO_VIRTUAL_MACHINE, proto.Provider_HOSTING).
```
Comment (chlam4) [inline prometurbo/pkg/registration/supply_chain_factory.go]: With this supply chain we are only supporting the case in which `hostedOnVM` is true, right?  If so, then I'd just suggest to add a TODO/comment here to call that out.  Just a nit though.

File: appmetric/pkg/config/appmetric-config.yaml [hunk 4]
```diff
@@ -118,6 +114,34 @@ exporters:
             isIdentifier: true
           target:
             label: job
+  - name: jmx-xl
+    entities:
+      - type: APPLICATION
+        metrics:
+          TRANSACTION:
+            queries:
+              used: 'sum by (instance,service) (delta(grpc_server_handled_total{job="xl",code="OK"}[5m]) or delta(api_call_counts{job="xl",failed="false"}[5m]))/300'
```
[the remark this answers is not in the document]
Comment (ading1977) [reply appmetric/pkg/config/appmetric-config.yaml]: Added a TODO comment in the yaml file

File: prometurbo/pkg/discovery/constant/constants.go [hunk 5]
```diff
@@ -17,23 +17,42 @@ const (
...
+var SupportedVMCommodities = CommodityTypeMap{
+	proto.CommodityDTO_VCPU: {},
 	proto.CommodityDTO_VMEM: {},
 }
 
-var AppCommodityTypeMap = map[proto.CommodityDTO_CommodityType]struct{}{
-	proto.CommodityDTO_TRANSACTION:   {},
-	proto.CommodityDTO_RESPONSE_TIME: {},
+var SupportedAppCommodities = CommodityTypeMap{
+	proto.CommodityDTO_TRANSACTION:     {},
```
[the remark this answers is not in the document]
Comment (ading1977) [reply prometurbo/pkg/discovery/constant/constants.go]: Yes, if no default, the capacity is 0, which is ignored, meaning, the capacity will not be set for the commodity (unless there is a query for capacity defined)

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go (hunk 1 above)
Comment (ading1977) [reply prometurbo/pkg/discovery/dtofactory/entity_builder.go]: This is to check if the type of the commodity obtained from the input metric is supported for the specific entity. If not, we will ignore this metric.

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go (hunk 2 above)
Comment (ading1977) [reply prometurbo/pkg/discovery/dtofactory/entity_builder.go]: Used value is always there. If it is not defined in the yaml, the validation of the yaml will fail:

```
func newMetricDef(metricConfig config.MetricConfig) (*metricDef, error) {
	if len(metricConfig.Queries) == 0 {
		return nil, fmt.Errorf("empty queries")
	}
	if _, exist := metricConfig.Queries[Used]; !exist {
		return nil, fmt.Errorf("missing query for used value")
	}
	metricDef := metricDef{
		queries: make(map[string]string),
	}
	for k, v := range metricConfig.Queries {
		metricDef.queries[k] = v
	}
	return &metricDef, nil
}
```

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go [hunk 6]
```diff
@@ -173,94 +151,79 @@ func (b *entityBuilder) createProviderEntity(ip string) (*proto.EntityDTO, error
...
+	for _, comm := range commodities {
+		commTypes = append(commTypes, *comm.CommodityType)
 	}
+	provider := builder.CreateProvider(entityType, providerId)
+	vAppType := proto.EntityDTO_VIRTUAL_APPLICATION
+	id := b.getEntityId(vAppType, ip)
+	vappDto, err := builder.NewEntityDTOBuilder(vAppType, id).
+		DisplayName(id).
+		Provider(provider).
+		BuysCommodities(commodities).
```
[the remark this answers is not in the document]
Comment (ading1977) [reply prometurbo/pkg/discovery/dtofactory/entity_builder.go]: Yes, for each provider and its corresponding commodities, we need to make sure to call `Provider` first, and then `BuysCommodities` next, then the go-sdk will properly fill the commoditiesBought map with the provider as the key, and the list of commodities as the value.

File: prometurbo/pkg/discovery/dtofactory/entity_builder.go [hunk 7]
```diff
@@ -173,94 +151,79 @@ func (b *entityBuilder) createProviderEntity(ip string) (*proto.EntityDTO, error
...
-		commodity, err := builder.NewCommodityDTOBuilder(commType).Used(0).Create()
+	commodities := providerDTO.CommoditiesSold
 
+	if b.metric.HostedOnVM {
+		if entityType != proto.EntityDTO_APPLICATION && entityType != proto.EntityDTO_DATABASE_SERVER {
+			return nil, fmt.Errorf("unsupported provider type %v to create consumer", entityType)
+		}
+		// Hosted on VM, create non-proxy Virtual Application entity
+		provider := builder.CreateProvider(entityType, providerId)
+		vAppType := proto.EntityDTO_VIRTUAL_APPLICATION
```
[the remark this answers is not in the document]
Comment (ading1977) [reply prometurbo/pkg/discovery/dtofactory/entity_builder.go]: This will be posted in a new PR.

**Pull request merged**